usb/ehci/ehci.c

```diff
@@ -36,166 +36,271 @@
 
 #define EHCI_PERIODIC_SIZE 128
 
-struct qtd_node {
-	struct qtd_node *prev, *next;
-	struct qtd *qtd;
-};
-
-
-struct qh_node {
-	struct qh_node *prev, *next;
-	struct qtd *last;
-	struct qh *qh;
-	unsigned period; /* [ms], interrupt transfer only */
-	unsigned phase;  /* [ms], interrupt transfer only */
-	unsigned uframe; /* interrupt transfer and high-speed only */
-};
-
 
 static inline void ehci_memDmb(void)
 {
 	asm volatile ("dmb" ::: "memory");
 }
 
 
-static int ehci_linkQtd(struct qtd *prev, struct qtd *next)
+static void ehci_startAsync(hcd_t *hcd)
 {
-	prev->next.pointer = va2pa(next) >> 5;
-	prev->next.type = 0;
-	prev->next.terminate = 0;
+	ehci_t *ehci = (ehci_t *)hcd->priv;
 
-	return 0;
+	*(hcd->base + asynclistaddr) = va2pa((void *)ehci->asyncList->hw);
+	*(hcd->base + usbcmd) |= USBCMD_ASE;
+	while ((*(hcd->base + usbsts) & USBSTS_AS) == 0)
+		;
+}
+
+
+static void ehci_stopAsync(hcd_t *hcd)
+{
+	*(hcd->base + usbcmd) &= ~USBCMD_ASE;
+	while ((*(hcd->base + usbsts) & USBSTS_AS) != 0)
```

Comment:
Memory barier before while would be a good idea

Reply:
Yes, I agree.